### PR TITLE
Scripted keyword field

### PR DESCRIPTION
**Pull request by javanna**
This is the first draft implementation of scripted field of type keyword. Still lots of TODOs to address, this PR targets a feature branch.

Relates to #59332

### CODE REVIEW

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/AbstractScriptFieldScript.java [hunk 1]
```diff
@@ -21,23 +24,19 @@
  */
 public abstract class AbstractScriptFieldScript {
     private final Map<String, Object> params;
-    private final LeafReaderContext ctx;
-    private final SourceLookup source;
-    private final LeafDocLookup fieldData;
+    private final LeafSearchLookup leafSearchLookup;
```
Comment (nik9000) [inline x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/AbstractScriptFieldScript.java:27]: Ok with me.

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/AbstractScriptFieldScript.java [hunk 2]
```diff
@@ -21,23 +24,19 @@
...
 
-    public AbstractScriptFieldScript(Map<String, Object> params, SourceLookup source, DocLookup fieldData, LeafReaderContext ctx) {
+    public AbstractScriptFieldScript(Map<String, Object> params, SearchLookup searchLookup, LeafReaderContext ctx) {
+        this.leafSearchLookup = searchLookup.getLeafSearchLookup(ctx);
+        //TODO how do other scripts get stored fields exposed? Through asMap? I don't see any getters for them.
```
Comment (nik9000) [inline x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/AbstractScriptFieldScript.java:31]: Yeah, they have them exposed in *weird* ways that I don't feel comfortable copying. I'd prefer to make a getter for them.

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/RuntimeFields.java [hunk 3]
```diff
@@ -6,25 +6,54 @@
...
+
+    @Override
+    public Collection<Object> createComponents(Client client, ClusterService clusterService, ThreadPool threadPool,
+                                               ResourceWatcherService resourceWatcherService, ScriptService scriptService,
+                                               NamedXContentRegistry xContentRegistry, Environment environment,
+                                               NodeEnvironment nodeEnvironment, NamedWriteableRegistry namedWriteableRegistry,
+                                               IndexNameExpressionResolver indexNameExpressionResolver,
+                                               Supplier<RepositoriesService> repositoriesServiceSupplier) {
+        //looks like createComponents gets called after getMappers
+        this.scriptTypeParser.setScriptService(scriptService);
```
Comment (nik9000) [inline x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/RuntimeFields.java:56]: That is kind of nasty. Not a blocker for the PR, and maybe not a blocker ever, but a little sad!

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/RuntimeKeywordMappedFieldType.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        return new RuntimeKeywordMappedFieldType(this);
+    }
+
+    @Override
+    public Object valueForDisplay(Object value) {
+        if (value == null) {
+            return null;
+        }
+        // keywords are internally stored as utf8 bytes
+        BytesRef binaryValue = (BytesRef) value;
```
Comment (nik9000) [inline x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/RuntimeKeywordMappedFieldType.java:52]: I think our keywords come out as strings though, right? When does this get called?

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/ScriptFieldMapper.java [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+    private static final FieldType FIELD_TYPE = new FieldType();
+
+    ScriptFieldMapper(String simpleName, MappedFieldType mappedFieldType, MultiFields multiFields, CopyTo copyTo) {
+        super(simpleName, FIELD_TYPE, mappedFieldType, multiFields, copyTo);
+    }
+
+    @Override
+    protected void parseCreateField(ParseContext context) {
+        //there is no field!
```
Comment (nik9000) [inline x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/ScriptFieldMapper.java:40]: This'll get called if the source contains the field. I wonder if we should do anything about that? Like, not now, but maybe sometime? I dunno.

**Review by nik9000 — APPROVED** [review]
LGTM, TODOs and al.

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/RuntimeFields.java (hunk 3 above)
Comment (javanna) [reply x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/RuntimeFields.java:56]: yea this caught me by surprise. Though it was not hard to work around.

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/RuntimeKeywordMappedFieldType.java (hunk 4 above)
Comment (javanna) [reply x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/RuntimeKeywordMappedFieldType.java:52]: no idea, this is copied from the keyword field....

**Comment by javanna** [comment]
@romseygeek given all you are doing with mappings, would you like to have a look too and give us some feedback?

**Pull request merged**